### PR TITLE
Replace Travis CI with GHA

### PR DESCRIPTION
**Pull request by JohnTitor**
On the rust-lang org, we've deprecated the use of Travis CI and migrated to GHA, this PR does it for this repo.
Tested the gh-pages works fine on my fork: https://johntitor.github.io/types-team/
Note that I don't use the scripts on simpleinfra to deploy the gh-pages, as it's outdated and not recommended to use (I remember Pietro said it before, but cannot find the comment on GitHub...)

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
Hm, did we repurpose a repo or something? Where did the Travis file come from - is there a template that needs adjusting?

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+
+jobs:
+  book-test:
+    runs-on: ubuntu-20.04
```
Comment (jyn514) [inline .github/workflows/main.yml:11]:
```suggestion
    runs-on: ubuntu-22.04
```

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+          git config user.name "Deploy from CI"
+          git config user.email ""
+          cd gh-pages
+          # Delete the ref to avoid keeping history.
+          git update-ref -d refs/heads/gh-pages
+          rm -rf *
+          mv ../book/* .
+          git add .
+          git commit -m "Deploy $GITHUB_SHA to gh-pages"
+          git push --force
```
Comment (jyn514) [inline .github/workflows/main.yml:43]: is this copied from somewhere? There's a lot git munging here, are you sure it works?

**Comment by jackh726** [comment]
@Mark-Simulacrum This repo got renamed wg-traits->types-team

**Comment by JohnTitor** [comment]
> Hm, did we repurpose a repo or something? Where did the Travis file come from - is there a template that needs adjusting?

Looking at https://github.com/rust-lang/types-team/commit/2c992aecfed391e416e402e97bb6a4d04dbfee30, the reason would be "we weren't ready to use GHA yet".

File: .github/workflows/main.yml (hunk 2 above)
Comment (JohnTitor) [reply .github/workflows/main.yml:43]: Yeah, I use it on libc crate: https://github.com/rust-lang/libc/blob/master/.github/workflows/docs.yml
And as I said on OP, it works fine on my fork: https://johntitor.github.io/types-team/

**Review by jyn514 — APPROVED** [review]

**Comment by jyn514** [comment]
> Note that I don't use the scripts on simpleinfra to deploy the gh-pages, as it's outdated and not recommended to use (I remember Pietro said it before, but cannot find the comment on GitHub...)

Can we update those scripts to use the code you have here so people can start using them again? That way we don't have to copy-paste these into each repo.

**Comment by JohnTitor** [comment]
> Can we update those scripts to use the code you have here so people can start using them again? That way we don't have to copy-paste these into each repo.

IIRC the reason that it's not recommended is that action is based on Docker (@pietroalbini Am I correct?): https://github.com/rust-lang/simpleinfra/blob/master/github-actions/static-websites/Dockerfile
It's generally slower than JavaScript's so it isn't straightforward, I guess.

**Comment by pietroalbini** [comment]
Yeah, the existing Docker action is painfully slow. We should figure out a better solution, but in the meantime the code looks correct!

**Comment by JohnTitor** [comment]
Could anyone merge this?

**Pull request merged**